### ISSUE TITLE
Crash Log: ElasticSearch EOF - index out of range

### ISSUE THREAD

**Issue by jvasallo** [terraform-providers/terraform-provider-aws#2138]
### Terraform Version
[jvasallo@jvasallo-mbp2016 us-east-1]$ terraform --version
Terraform v0.10.8

### Affected Resource(s)
- aws_elasticsearch_domain

### Terraform Configuration Files
```hcl
resource "aws_elasticsearch_domain" "demo-es" {
  domain_name           = "demo-es"
  elasticsearch_version = "5.5"

  cluster_config {
    instance_type = "t2.medium.elasticsearch"
  }

  advanced_options {
    "rest.action.multi.allow_explicit_index" = "true"
  }

  ebs_options {
    ebs_enabled = true
    volume_size = 35
  }

  access_policies = <<CONFIG
  {
    "Version": "2012-10-17",
    "Statement": [
      {
        "Effect": "Allow",
        "Principal": {
          "AWS": "*"
        },
        "Action": "es:*",
        "Resource": "arn:aws:es:us-east-1:${var.account_number}:domain/demo-es/*"
      }
    ]
  }
  CONFIG

  tags {
    Name      = "demo-es"
    app_group = "demoapps"
    app_name  = "demotestapps"
  }
}
```

### Debug Output
Available upon request.

### Crash Output
https://gist.github.com/jvasallo/b127e2fd6cd674cb3c945ab77a1b016f

### Expected Behavior
Terraform to apply the proper policy

### Actual Behavior
Terraform crashes with EOF error.

### Steps to Reproduce
1. `terraform apply`

### Important Factoids
Running on Mac OS High Sierra, `terraform apply` works fine for another coworker running Fedora 26. Not sure if that helps.

### References
- N/A (To my knowledge)

**Comment by bflad** [comment]
It looks like the above Terraform configuration may not follow the correct heredoc syntax as described in the [multiline string documentation](https://www.terraform.io/docs/configuration/syntax.html). Try removing the indentation/whitespace before `CONFIG` heredoc ending.

Please also note you _may_ need to remove the whitespace before the opening `{` of your JSON as well, some resources or AWS itself may throw the JSON validation error of "expected first character to be {"

If the above fixes your error, maybe this issue should be migrated to Terraform core/HCL for triage or additional documentation.

**Comment by jvasallo** [comment]
## Resource Config
```
resource "aws_elasticsearch_domain" "demo-es" {
  domain_name           = "demo-es"
  elasticsearch_version = "5.5"

  cluster_config {
    instance_type = "t2.medium.elasticsearch"
  }

  advanced_options {
    "rest.action.multi.allow_explicit_index" = "true"
  }

  ebs_options {
    ebs_enabled = true
    volume_size = 35
  }

  access_policies = <<CONFIG
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "AWS": "*"
            },
            "Action": "es:*",
            "Resource": "arn:aws:es:us-east-1:${var.account_number}:domain/demo-es/*"
        }
    ]
}
CONFIG

  tags {
    Name      = "demo-es"
    app_group = "demoapps"
    app_name  = "demotestapps"
  }
}
```

## Steps:
- Ran `terraform fmt`
- Ran in vim `:set list`
- Manually spaced the files just in case
- Works in Linux (Fedora 26 and Ubuntu) but not on Mac (High Sierra)

Still seeing the error unfortunately. Any other issues you can think of? Following the doc here:
https://www.terraform.io/docs/providers/aws/r/elasticsearch_domain.html

**Comment by jvasallo** [comment]
Just found out why it was working in Linux, it appears we were running an older version of the provider: `terraform-provider-aws_v0.1.1_x4`

I am running: `terraform-provider-aws_v1.2.0_x4`

After upgrading via `terraform init -upgrade`, it began causing issues on Linux as well. At this time, my issue is no longer related to OS (very confused if that was the case), but rather the terraform provider version.

**Comment by jvasallo** [comment]
I was able to get my `terraform apply` to succeed after downgrading the provider to an earlier version. 

Here is non working pair:
```
- Downloading plugin for provider "template" (1.0.0)...
- Downloading plugin for provider "aws" (1.2.0)...
```

Here is the working combo:
```
- Downloading plugin for provider "aws" (1.1.0)...
- Downloading plugin for provider "template" (1.0.0)...
```

**Comment by radeksimko** [comment]
Hi @jvasallo 
thanks for the report.

Based on the crash log it looks like this was caused by removing the `vpc_config` block (which was in the config previously) and trying to "update" the ES Domain.

I'm pretty confident this was addressed in https://github.com/terraform-providers/terraform-provider-aws/pull/2157 which was released as part of [1.3.0](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#130-november-16-2017). Turning VPC-based ES domain into non-VPC and vice-versa is not something that can be done during update per [documentation](https://docs.aws.amazon.com/elasticsearch-service/latest/developerguide/es-vpc.html#es-vpc-limitations).

> If you launch a new domain within a VPC, you can't later switch it to use a public endpoint. The reverse is also true: If you create a domain with a public endpoint, you can't later place it within a VPC. Instead, you must create a new domain and migrate your data.

which is why `plan` will now (from 1.3.0 onwards) present recreation of the domain in case user decides to remove `vpc_config`.

Let me know if the problem still exists after upgrading to `1.3.0+`. You can use `terraform init -upgrade` to upgrade the provider.
